phd-tests/framework/src/zfs.rs

```diff
@@ -0,0 +1,321 @@
+    /// operation.
+    name: CloneName,
+
+    /// The point at which this clone is mounted in the global directory
+    /// hierarchy.
+    mount_point: Utf8PathBuf,
+
+    /// The snapshot this clone derives from. Snapshots can't be deleted until
+    /// all their clones are gone; this reference helps to ensure that clones
+    /// and snapshots are deleted in the correct order irrespective of when the
+    /// clones and the [`CloneManager`] are dropped.
+    _snapshot: Snapshot,
+}
+
+impl Drop for Clone {
+    fn drop(&mut self) {
+        debug!(name = self.name.0, "zfs clone dropped");
+        let _ = zfs_command("destroy", &[&self.name.0]);
+    }
+}
+
+/// Represents a specific copy-on-write file within a ZFS clone. When this is
+/// dropped, attempts to delete the associated clone.
+#[derive(Debug)]
+pub struct ClonedFile {
+    /// The clone to which this file belongs.
+    clone: Clone,
+
+    /// The path to this file relative to the mount point of the clone.
+    relative_path: Utf8PathBuf,
+}
+
+impl ClonedFile {
+    /// Creates a snapshot and clone of the dataset that contains the canonical
+    /// location of the file indicated by `path`.
+    pub fn create_from_path(path: &Utf8Path) -> anyhow::Result<Self> {
+        // Canonicalize the path to resolve any symbolic links before doing any
+        // prefix matching.
+        let canonical_path = path.canonicalize_utf8()?;
+
+        let containing_dataset = Dataset::from_path(&canonical_path)
+            .with_context(|| format!("getting dataset containing {path}"))?;
+
+        let relative_file_path = canonical_path
+            .strip_prefix(&containing_dataset.mount_point)
+            .context("getting relative path to file to clone")?;
+
+        let snapshot = Snapshot::create_from_dataset(&containing_dataset.name)?;
+        Self::create_from_paths_and_snapshot(
+            containing_dataset,
+            relative_file_path,
+            snapshot,
+        )
+        .context("creating zfs clone")
+    }
+
+    /// Yields the absolute path to this cloned file in the global directory
+    /// hierarchy.
+    pub fn path(&self) -> Utf8PathBuf {
+        let mut path = self.clone.mount_point.clone();
+        path.push(&self.relative_path);
+        path
+    }
+
+    /// Given a path to a file relative to the root of its (mounted) dataset,
+    /// and the name of a snapshot of that dataset, clones the snapshot and
+    /// returns a handle to the clone. The [`path`] method can be used to find
+    /// the absolute path to the file within the clone.
+    fn create_from_paths_and_snapshot(
+        dataset: Dataset,
+        relative_file_path: &Utf8Path,
+        snapshot: Snapshot,
+    ) -> anyhow::Result<Self> {
+        let clone_name =
+            format!("{}/phd-clone-{}", dataset.name.0, Uuid::new_v4());
+
+        zfs_command("clone", &[&snapshot.name.0, &clone_name])?;
+
+        fn get_clone_mount_point(
```

Comment:
maybe worth a comment explaining that this is a function so that we can run `zfs destroy` no matter where we bail in this function?

Reply:
Done in dece78c.

Reply:
Added a note in dece78ce.